grimoirelab_toolkit/credential_manager/bw_manager.py

```diff
@@ -0,0 +1,421 @@
+        self.formatted_credentials = {}
+        # store email for session validation
+        self._email = email
+        self.last_sync_time = None
+        self.sync_interval = timedelta(minutes=3)
+
+        try:
+            self._login(email, password)
+        except (InvalidCredentialsError, BitwardenCLIError, ConnectionError) as e:
+            logger.error("Failed to initialize Bitwarden manager: %s", e)
+            raise
+
+    def _login(self, bw_email: str, bw_password: str) -> str | None:
+        """
+        Logs into Bitwarden and obtains a session key.
+
+        Method follows these steps:
+        1. Check if current session is valid and return if true
+        2. Get current Bitwarden status
+        3. Handle authentication for already logged user or not logged user
+        4. Sync vault if needed
+
+        :param str bw_email: Bitwarden account email.
+        :param str bw_password: Bitwarden account password.
+        :returns: The session key for the current Bitwarden session.
+        :rtype: str
+        :raises InvalidCredentialsError: If invalid credentials are provided
+        :raises SessionNotFoundError: If session cannot be established
+        :raises BitwardenCLIError: If Bitwarden CLI operations fail
+        :raises CredentialConnectionError: If connection issues occur
+        """
+        try:
+            if self._is_session_valid():
+                self._sync_if_needed()
+                return self.session_key
+
+            status = self._get_bitwarden_status()
+
+            if not status:
+                raise SessionNotFoundError("Could not retrieve Bitwarden status")
+
+            if status.get("userEmail") == bw_email:
+                # the email in the returned status is the same introduced, so the user is logged in
+                self._handle_existing_user_auth(bw_password, status)
+            else:
+                # logs in with the desired credentials
+                self._handle_new_user_auth(bw_email, bw_password)
+
+            if not self.session_key:
+                # there is no session key after atempting to log in
+                raise SessionNotFoundError("Session key not found cause could not log in")
+
+            self._sync_if_needed()
+            return self.session_key
+
+        except (subprocess.CalledProcessError, json.JSONDecodeError) as e:
+            raise ConnectionError(f"Bitwarden CLI command failed: {e}")
+
+    def _is_session_valid(self) -> bool:
+        """
+        Check if there is an active session and doesn't need re-authentication.
+
+        :returns: True if session is valid and active, False otherwise
+        :rtype: bool
+        """
+        if self.session_key is not None and self._validate_session():
+            return True
+        return False
+
+    def _sync_if_needed(self) -> None:
+        """
+        Sync vault if sync interval has passed.
+
+        """
+        if self._should_sync():
+            self._sync_vault()
```

Comment:
I think you just need one method called `sync_vault` that should do the check and if it fails, synchronize the vault. Take also into account that `_should_sync()` could just be in `_sync_if_needed()` because it's just only asserting the context.